File: 733-flood-fill/733-flood-fill.scala
import scala.collection.mutable._

object Solution {
    def floodFill(image: Array[Array[Int]], sr: Int, sc: Int, newColor: Int): Array[Array[Int]] = {
        val c = image(sr)(sc)
        var q: List[Array[Int]] = Nil
        var next = (0, 1) :: (0, -1) :: (1, 0) :: (-1, 0) :: Nil
        
        if (c != newColor) {
            q = q :+ Array(sr, sc)
        }
        
        while(q.nonEmpty) {
            val p :: rest = q
            q = rest
            
            image(p(0))(p(1)) = newColor
            
            next.foreach { np =>
                val nr = p(0) + np._1
                val nc = p(1) + np._2
                if (nr >= 0 && nr < image.length && nc >= 0 && nc < image(0).length && image(nr)(nc) == c) {
                    q = q :+ Array(nr, nc)
                }
            }
        }
        
        image
    }
}

File: 733-flood-fill/733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int oriClr = image[sr][sc];
        
        if (oriClr == color) return image;
        
        int[][] next = new int[][] {
            {0, 1},
            {0, -1},
            {1, 0},
            {-1, 0}
        };
        
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[] {sr, sc});
        
        while(!queue.isEmpty()) {
            int[] loc = queue.poll();
            
            image[loc[0]][loc[1]] = color;
            
            for(int[] n : next) {
                int nr = loc[0] + n[0];
                int nc = loc[1] + n[1];
                
                if (nr < 0 || nc < 0 || nr >= image.length || nc >= image[0].length || image[nr][nc] != oriClr) {
                    continue;
                }
                
                queue.offer(new int[] {nr, nc});
            }
        }
        
        return image;
    }
}